arch/x86/crc32_fold_pclmulqdq.c

```diff
@@ -17,19 +17,20 @@
  */
 
 #ifdef X86_PCLMULQDQ_CRC
+#include "../../zutil.h"
 
 #include <inttypes.h>
 #include <immintrin.h>
 #include <wmmintrin.h>
 
-#include "crc_folding.h"
+#include "../../crc32_fold.h"
 
-Z_INTERNAL uint32_t crc_fold_init(unsigned int crc0[4 * 5]) {
+Z_INTERNAL uint32_t crc32_fold_reset_pclmulqdq(crc32_fold *crc) {
     /* CRC_SAVE */
-    _mm_storeu_si128((__m128i *)crc0 + 0, _mm_cvtsi32_si128(0x9db42487));
-    _mm_storeu_si128((__m128i *)crc0 + 1, _mm_setzero_si128());
-    _mm_storeu_si128((__m128i *)crc0 + 2, _mm_setzero_si128());
-    _mm_storeu_si128((__m128i *)crc0 + 3, _mm_setzero_si128());
+    _mm_storeu_si128((__m128i *)crc->fold + 0, _mm_cvtsi32_si128(0x9db42487));
+    _mm_storeu_si128((__m128i *)crc->fold + 1, _mm_setzero_si128());
+    _mm_storeu_si128((__m128i *)crc->fold + 2, _mm_setzero_si128());
+    _mm_storeu_si128((__m128i *)crc->fold + 3, _mm_setzero_si128());
```

Comment:
Using temporary variable would make this cleaner as it would remove need for repeated casting.
Using unaligned version should also be unnecessary as the structure is guaranteed to be aligned and fold is the first member of the structure.

Reply:
Fixed.